File: windows/QMK Toolbox/USB.cs
using System;
using System.Linq;
using System.Management;
using System.Text.RegularExpressions;

namespace QMK_Toolbox
{
    public class Usb
    {
        private readonly int[] _devicesAvailable = new int[(int)Chipset.NumberOfChipsets];
        private readonly Flashing _flasher;
        private readonly Printing _printer;
        private readonly Regex DeviceIdRegex = new Regex(@"USB\\VID_([0-9A-F]+).*PID_([0-9A-F]+).*REV_([0-9A-F]+).*");

        private ManagementEventWatcher deviceConnectedWatcher;
        private ManagementEventWatcher deviceDisconnectedWatcher;

        public Usb(Flashing flasher, Printing printer)
        {
            _flasher = flasher;
            _printer = printer;
        }

        public  void StartListeningForDeviceEvents(EventArrivedEventHandler handler)
        {
            if (deviceConnectedWatcher == null)
            {
                deviceConnectedWatcher = StartManagementEventWatcher("__InstanceCreationEvent", handler);
            }

            if (deviceDisconnectedWatcher == null)
            {
                deviceDisconnectedWatcher = StartManagementEventWatcher("__InstanceDeletionEvent", handler);
            }
        }

        public void StopListeningForDeviceEvents()
        {
            if (deviceConnectedWatcher != null)
            {
                deviceConnectedWatcher.Stop();
            }

            if (deviceDisconnectedWatcher != null)
            {
                deviceDisconnectedWatcher.Stop();
            }
        }

        private ManagementEventWatcher StartManagementEventWatcher(string eventType, EventArrivedEventHandler handler)
        {
            var watcher = new ManagementEventWatcher($"SELECT * FROM {eventType} WITHIN 2 WHERE TargetInstance ISA 'Win32_PnPEntity' AND TargetInstance.DeviceID LIKE 'USB%'");
            watcher.EventArrived += handler;
            watcher.Start();
            return watcher;
        }

        public bool DetectBootloaderFromCollection(ManagementObjectCollection collection, bool connected = true)
        {
            var found = false;
            foreach (var instance in collection)
            {
                found = DetectBootloader(instance, connected);
            }
            return found;
        }

        private static readonly ushort[] caterinaVids =
        {
            0x1B4F, // Spark Fun Electronics
            0x1FFB, // Pololu Electronics
            0x2341, // Arduino SA
            0x239A, // Adafruit Industries LLC
            0x2A03  // dog hunter AG
        };

        private static readonly ushort[] caterinaPids =
        {
            // Adafruit Industries LLC
            0x000C, // Feather 32U4
            0x000D, // ItsyBitsy 32U4 3V3/8MHz
            0x000E, // ItsyBitsy 32U4 5V/16MHz
            // Arduino SA / dog hunter AG
            0x0036, // Leonardo
            0x0037, // Micro
            // Pololu Electronics
            0x0101, // A-Star 32U4
            // Spark Fun Electronics
            0x9203, // Pro Micro 3V3/8MHz
            0x9205, // Pro Micro 5V/16MHz
            0x9207  // LilyPad 3V3/8MHz (and some Pro Micro clones)
        };

        private static readonly ushort[] atmelDfuPids =
        {
            0x2FEF, // ATmega16U2
            0x2FF0, // ATmega32U2
            0x2FF3, // ATmega16U4
            0x2FF4, // ATmega32U4
            0x2FF9, // AT90USB64
            0x2FFA, // AT90USB162
            0x2FFB  // AT90USB128
        };

        public bool DetectBootloader(ManagementBaseObject instance, bool connected = true)
        {
            var deviceId = GetHardwareId(instance);

            var vpr = DeviceIdRegex.Match(deviceId);
            if (vpr.Success)
            {
                var vendorId = Convert.ToUInt16(vpr.Groups[1].ToString(), 16);
                var productId = Convert.ToUInt16(vpr.Groups[2].ToString(), 16);
                var revisionBcd = Convert.ToUInt16(vpr.Groups[3].ToString(), 16);

                string deviceName;
                string comPort = null;
                string mountPoint = null;
                string driverName = GetDriverName(instance);
                Chipset deviceType;

                if (IsSerialDevice(instance))
                {
                    if (vendorId == 0x03EB && productId == 0x6124) // Atmel SAM-BA
                    {
                        deviceName = "Atmel SAM-BA";
                        deviceType = Chipset.AtmelSamBa;
                    }
                    else if (caterinaVids.Contains(vendorId) && caterinaPids.Contains(productId)) // Caterina
                    {
                        deviceName = "Caterina";
                        deviceType = Chipset.Caterina;
                    }
                    else if (vendorId == 0x16C0 && productId == 0x0483) // ArduinoISP/AVRISP
                    {
                        deviceName = "AVRISP";
                        deviceType = Chipset.AvrIsp;
                    }
                    else
                    {
                        return false;
                    }

                    comPort = GetComPort(instance);
                    _flasher.ComPort = comPort;
                }
                else if (vendorId == 0x03EB)
                {
                    if (atmelDfuPids.Contains(productId))
                    {
                        if (revisionBcd == 0x0936) // QMK-DFU
                        {
                            deviceName = "QMK DFU";
                            deviceType = Chipset.QmkDfu;
                        }
                        else // Atmel DFU
                        {
                            deviceName = "Atmel DFU";
                            deviceType = Chipset.AtmelDfu;
                        }
                    }
                    else if (productId == 0x2045) // LUFA MS
                    {
                        deviceName = "LUFA Mass Storage";
                        deviceType = Chipset.LufaMs;
                        mountPoint = GetMountPoint(instance);
                        _flasher.MountPoint = mountPoint;
                    }
                    else
                    {
                        return false;
                    }
                }
                else if (vendorId == 0x16C0 && productId == 0x0478) // PJRC Teensy
                {
                    deviceName = "Halfkay";
                    deviceType = Chipset.Halfkay;
                }
                else if (vendorId == 0x0483 && productId == 0xDF11) // STM32 DFU
                {
                    deviceName = "STM32 DFU";
                    deviceType = Chipset.Stm32Dfu;
                }
                else if (vendorId == 0x314B && productId == 0x0106) // APM32 DFU
                {
                    deviceName = "APM32 DFU";
                    deviceType = Chipset.Apm32Dfu;
                }
                else if (vendorId == 0x1C11 && productId == 0xB007) // Kiibohd
                {
                    deviceName = "Kiibohd";
                    deviceType = Chipset.Kiibohd;
                }
                else if (vendorId == 0x16C0 && productId == 0x05DF) // Objective Development BootloadHID
                {
                    deviceName = "BootloadHID";
                    deviceType = Chipset.BootloadHid;
                }
                else if (vendorId == 0x16C0 && productId == 0x05DC) // USBasp and USBaspLoader
                {
                    deviceName = "USBasp";
                    deviceType = Chipset.UsbAsp;
                }
                else if (vendorId == 0x1781 && productId == 0x0C9F) // AVR Pocket ISP
                {
                    deviceName = "USB Tiny";
                    deviceType = Chipset.UsbTiny;
                }
                else if (vendorId == 0x1EAF && productId == 0x0003) // STM32Duino
                {
                    deviceName = "STM32Duino";
                    deviceType = Chipset.Stm32Duino;
                }
                else
                {
                    return false;
                }

                var connectedString = connected ? "connected" : "disconnected";
                var comPortString = comPort != null ? $" [{comPort}]" : "";
                var mountPointString = mountPoint != null ? $" [{mountPoint}]" : "";
                var driverString = driverName ?? "NO DRIVER";

                _printer.Print($"{deviceName} device {connectedString} ({driverString}): {instance.GetPropertyValue("Manufacturer")} {instance.GetPropertyValue("Name")} ({vendorId:X4}:{productId:X4}:{revisionBcd:X4}){comPortString}{mountPointString}", MessageType.Bootloader);

                _devicesAvailable[(int)deviceType] += (connected ? 1 : -1);

                return true;
            }

            return false;
        }

        public string GetHardwareId(ManagementBaseObject instance)
        {
            var hardwareIds = (string[])instance.GetPropertyValue("HardwareID");
            if (hardwareIds != null && hardwareIds.Length > 0)
            {
                return hardwareIds[0];
            }

            return null;
        }

        public string GetDriverName(ManagementBaseObject instance)
        {
            var service = (string)instance.GetPropertyValue("Service");
            if (service != null && service.Length > 0)
            {
                return service;
            }

            return null;
        }

        public bool IsSerialDevice(ManagementBaseObject instance)
        {
            var compatibleIds = (string[])instance.GetPropertyValue("CompatibleID");
            return (compatibleIds != null && compatibleIds.Contains("USB\\Class_02&SubClass_02")); // CDC-ACM
        }

        public string GetComPort(ManagementBaseObject instance)
        {
            using (var searcher = new ManagementObjectSearcher("SELECT PNPDeviceID, DeviceID FROM Win32_SerialPort"))
            {
                foreach (var device in searcher.Get())
                {
                    if (device.GetPropertyValue("PNPDeviceID").ToString().Equals(instance.GetPropertyValue("PNPDeviceID").ToString()))
                    {
                        return device.GetPropertyValue("DeviceID").ToString();
                    }
                }
            }

            return null;
        }

        public string GetMountPoint(ManagementBaseObject instance)
        {
            foreach (ManagementObject usbHub in new ManagementObjectSearcher("SELECT * FROM Win32_USBHub").Get())
            {
                if (usbHub.GetPropertyValue("PNPDeviceID").ToString().Equals(instance.GetPropertyValue("PNPDeviceID").ToString()))
                {
                    foreach (ManagementObject usbController in usbHub.GetRelated("Win32_USBController"))
                    {
                        foreach (ManagementObject assoc in new ManagementObjectSearcher("ASSOCIATORS OF {Win32_USBController.DeviceID='" + usbController["PNPDeviceID"].ToString() + "'}").Get())
                        {
                            if (assoc.GetPropertyValue("CreationClassName").Equals("Win32_PnPEntity") && assoc.GetPropertyValue("DeviceID").ToString().Contains("USBSTOR"))
                            {
                                foreach (ManagementObject diskDrive in new ManagementObjectSearcher("SELECT * FROM Win32_DiskDrive").Get())
                                {
                                    if (diskDrive.GetPropertyValue("PNPDeviceID").ToString().Equals(assoc.GetPropertyValue("PNPDeviceID").ToString()))
                                    {
                                        foreach (ManagementObject partition in diskDrive.GetRelated("Win32_DiskPartition"))
                                        {
                                            foreach (ManagementObject logicalDisk in partition.GetRelated("Win32_LogicalDisk"))
                                            {
                                                return logicalDisk.GetPropertyValue("Name").ToString();
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }

            return null;
        }

        public bool CanFlash(Chipset chipset) => _devicesAvailable[(int)chipset] > 0;

        public bool AreDevicesAvailable()
        {
            var available = false;
            for (var i = 0; i < (int)Chipset.NumberOfChipsets; i++)
            {
                available |= _devicesAvailable[i] > 0;
            }
            return available;
        }
    }
}

File: macos/QMK Toolbox/USB.m
#import <IOKit/usb/IOUSBLib.h>
#import <IOKit/serial/IOSerialKeys.h>
#import <IOKit/storage/IOMedia.h>

#import "USB.h"

static io_iterator_t usbConnectedIter;
static io_iterator_t usbDisconnectedIter;

//Global variables
static IONotificationPortRef gNotifyPort;
static Printing *_printer;

NSArray *caterinaVids;
NSArray *caterinaPids;
NSArray *atmelDfuPids;

@interface USB () <USBDelegate>
@end

@implementation USB
static int devicesAvailable[NumberOfChipsets];

+ (void)setupWithPrinter:(Printing *)printer andDelegate:(id<USBDelegate>)d {
    delegate = d;
    _printer = printer;
    mach_port_t masterPort;
    CFRunLoopSourceRef runLoopSource;

    caterinaVids = @[
        @0x1B4F, // Spark Fun Electronics
        @0x1FFB, // Pololu Electronics
        @0x2341, // Arduino SA
        @0x239A, // Adafruit Industries LLC
        @0x2A03  // dog hunter AG
    ];
    caterinaPids = @[
        // Adafruit Industries LLC
        @0x000C, // Feather 32U4
        @0x000D, // ItsyBitsy 32U4 3V3/8MHz
        @0x000E, // ItsyBitsy 32U4 5V/16MHz
        // Arduino SA / dog hunter AG
        @0x0036, // Leonardo
        @0x0037, // Micro
        // Pololu Electronics
        @0x0101, // A-Star 32U4
        // Spark Fun Electronics
        @0x9203, // Pro Micro 3V3/8MHz
        @0x9205, // Pro Micro 5V/16MHz
        @0x9207  // LilyPad 3V3/8MHz (and some Pro Micro clones)
    ];
    atmelDfuPids = @[
        @0x2FEF, // ATmega16U2
        @0x2FF0, // ATmega32U2
        @0x2FF3, // ATmega16U4
        @0x2FF4, // ATmega32U4
        @0x2FF9, // AT90USB64
        @0x2FFA, // AT90USB162
        @0x2FFB  // AT90USB128
    ];

    //Create a master port for communication with the I/O Kit
    IOMasterPort(MACH_PORT_NULL, &masterPort);

    gNotifyPort = IONotificationPortCreate(masterPort);
    runLoopSource = IONotificationPortGetRunLoopSource(gNotifyPort);
    CFRunLoopAddSource(CFRunLoopGetCurrent(), runLoopSource, kCFRunLoopDefaultMode);

    CFMutableDictionaryRef usbMatcher = IOServiceMatching(kIOUSBDeviceClassName);
    usbMatcher = (CFMutableDictionaryRef) CFRetain(usbMatcher);

    IOServiceAddMatchingNotification(gNotifyPort, kIOFirstMatchNotification, usbMatcher, deviceConnectedEvent, NULL, &usbConnectedIter);
    deviceConnectedEvent(NULL, usbConnectedIter);

    IOServiceAddMatchingNotification(gNotifyPort, kIOTerminatedNotification, usbMatcher, deviceDisconnectedEvent, NULL, &usbDisconnectedIter);
    deviceDisconnectedEvent(NULL, usbDisconnectedIter);

    //Finished with master port
    mach_port_deallocate(mach_task_self(), masterPort);
    masterPort = 0;
}

+ (NSString *)stringProperty:(CFStringRef)property forDevice:(io_service_t)device {
    CFStringRef cfProperty = IORegistryEntryCreateCFProperty(device, property, kCFAllocatorDefault, kNilOptions);
    if (cfProperty != nil) {
        NSString *nsProperty = (__bridge NSString *)(cfProperty);
        CFRelease(cfProperty);
        return nsProperty;
    }
    return nil;
}

+ (NSString *)vendorStringForDevice:(io_service_t)device {
    return [USB stringProperty:CFSTR(kUSBVendorString) forDevice:device];
}

+ (NSString *)productStringForDevice:(io_service_t)device {
    return [USB stringProperty:CFSTR(kUSBProductString) forDevice:device];
}

+ (NSString *)calloutDeviceForDevice:(io_service_t)device {
    CFMutableDictionaryRef serialMatcher = IOServiceMatching(kIOSerialBSDServiceValue);
    CFDictionarySetValue(serialMatcher, CFSTR(kIOSerialBSDTypeKey), CFSTR(kIOSerialBSDAllTypes));

    NSNumber *childVendorID = [USB vendorIDForDevice:device];
    NSNumber *childProductID = [USB productIDForDevice:device];

    io_iterator_t serialIterator;
    IOServiceGetMatchingServices(kIOMasterPortDefault, serialMatcher, &serialIterator);

    io_service_t port;
    while ((port = IOIteratorNext(serialIterator))) {
        NSNumber *parentVendorID = [USB vendorIDForDevice:port];
        NSNumber *parentProductID = [USB productIDForDevice:port];

        if (parentVendorID != nil) {
            if ([parentVendorID isEqualTo:childVendorID] && [parentProductID isEqualTo:childProductID]) {
                return [USB stringProperty:CFSTR(kIOCalloutDeviceKey) forDevice:port];
            }
        }
    }
    return nil;
}

+ (NSString *)mountPointForDevice:(io_service_t)device {
    CFMutableDictionaryRef massStorageMatcher = IOServiceMatching(kIOMediaClass);
    CFDictionarySetValue(massStorageMatcher, CFSTR(kIOMediaRemovableKey), kCFBooleanTrue);

    NSNumber *childVendorID = [USB vendorIDForDevice:device];
    NSNumber *childProductID = [USB productIDForDevice:device];

    io_iterator_t massStorageIterator;
    IOServiceGetMatchingServices(kIOMasterPortDefault, massStorageMatcher, &massStorageIterator);

    io_service_t media;
    while ((media = IOIteratorNext(massStorageIterator))) {
        NSNumber *parentVendorID = [USB vendorIDForDevice:media];
        NSNumber *parentProductID = [USB productIDForDevice:media];

        if (parentVendorID != nil) {
            if ([parentVendorID isEqualTo:childVendorID] && [parentProductID isEqualTo:childProductID]) {
                DASessionRef sessionRef = DASessionCreate(kCFAllocatorDefault);
                if (sessionRef) {
                    DADiskRef diskRef = DADiskCreateFromIOMedia(kCFAllocatorDefault, sessionRef, media);
                    if (diskRef) {
                        CFDictionaryRef diskProperties = DADiskCopyDescription(diskRef);
                        if (diskProperties) {
                            NSURL *mountPointUrl = [(__bridge NSDictionary*)diskProperties objectForKey:(NSString *)kDADiskDescriptionVolumePathKey];
                            return [mountPointUrl path];
                        }
                    }
                }
            }
        }
    }
    IOObjectRelease(media);

    return nil;
}

+ (NSNumber *)numberProperty:(CFStringRef)property forDevice:(io_service_t)device {
    CFNumberRef cfProperty = IORegistryEntrySearchCFProperty(device, kIOServicePlane, property, kCFAllocatorDefault, kIORegistryIterateParents | kIORegistryIterateRecursively);
    if (cfProperty != nil) {
        NSNumber *nsProperty = (__bridge NSNumber *)(cfProperty);
        CFRelease(cfProperty);
        return nsProperty;
    }
    return nil;
}

+ (NSNumber *)vendorIDForDevice:(io_service_t)device {
    return [USB numberProperty:CFSTR(kUSBVendorID) forDevice:device];
}

+ (NSNumber *)productIDForDevice:(io_service_t)device {
    return [USB numberProperty:CFSTR(kUSBProductID) forDevice:device];
}

+ (NSNumber *)revisionBCDForDevice:(io_service_t)device {
    return [USB numberProperty:CFSTR(kUSBDeviceReleaseNumber) forDevice:device];
}

+ (BOOL)isSerialDevice:(io_service_t)device {
    return [[USB numberProperty:CFSTR(kUSBDeviceClass) forDevice:device] isEqualTo:[NSNumber numberWithInt:kUSBCommunicationClass]];
}

static void deviceConnectedEvent(void *refCon, io_iterator_t iterator) {
    io_service_t device;
    while ((device = IOIteratorNext(iterator))) {
        [USB deviceEvent:device connected:YES];
    }
}

static void deviceDisconnectedEvent(void *refCon, io_iterator_t iterator) {
    io_service_t device;
    while ((device = IOIteratorNext(iterator))) {
        [USB deviceEvent:device connected:NO];
        IOObjectRelease(device);
    }
}

+ (void)deviceConnectedEvent:(io_service_t)device {
    [USB deviceEvent:device connected:YES];
}

+ (void)deviceDisconnectedEvent:(io_service_t)device {
    [USB deviceEvent:device connected:NO];
}

+ (void)deviceEvent:(io_service_t)device connected:(BOOL)connected {
    unsigned short vendorID = [[USB vendorIDForDevice:device] unsignedShortValue];
    unsigned short productID = [[USB productIDForDevice:device] unsignedShortValue];
    unsigned short revisionBCD = [[USB revisionBCDForDevice:device] unsignedShortValue];
    NSString *vendorString = [USB vendorStringForDevice:device];
    NSString *productString = [USB productStringForDevice:device];

    NSString *deviceName;
    NSString *calloutDevice;
    NSString *mountPoint;
    Chipset deviceType;

    if ([USB isSerialDevice:device]) {
        if (vendorID == 0x03EB && productID == 0x6124) { // Atmel SAM-BA
            deviceName = @"Atmel SAM-BA";
            deviceType = AtmelSAMBA;
        } else if ([caterinaVids containsObject:[NSNumber numberWithUnsignedShort:vendorID]] && [caterinaPids containsObject:[NSNumber numberWithUnsignedShort:productID]]) { // Caterina
            deviceName = @"Caterina";
            deviceType = Caterina;
        } else if (vendorID == 0x16C0 && productID == 0x0483) { // ArduinoISP/AVRISP
            deviceName = @"AVR-ISP";
            deviceType = AVRISP;
        } else {
            return;
        }

        if (connected) {
            while (calloutDevice == nil) {
                calloutDevice = [USB calloutDeviceForDevice:device];
            }
            [delegate setSerialPort:calloutDevice];
        }
    } else if (vendorID == 0x03EB) {
        if ([atmelDfuPids containsObject:[NSNumber numberWithUnsignedShort:productID]]) {
            if (revisionBCD == 0x0936) { // QMK DFU
                deviceName = @"QMK DFU";
                deviceType = QMKDFU;
            } else { // Atmel DFU
                deviceName = @"Atmel DFU";
                deviceType = AtmelDFU;
            }
        } else if (productID == 0x2045) { // LUFA MS
            deviceName = @"LUFA Mass Storage";
            deviceType = LUFAMS;

            if (connected) {
                while (mountPoint == nil) {
                    mountPoint = [USB mountPointForDevice:device];
                }
                [delegate setMountPoint:mountPoint];
            }
        } else {
            return;
        }
    } else if (vendorID == 0x16C0 && productID == 0x0478) { // PJRC Teensy
        deviceName = @"Halfkay";
        deviceType = Halfkay;
    } else if (vendorID == 0x0483 && productID == 0xDF11) { // STM32 DFU
        deviceName = @"STM32 DFU";
        deviceType = STM32DFU;
    } else if (vendorID == 0x314B && productID == 0x0106) { // APM32 DFU
        deviceName = @"APM32 DFU";
        deviceType = APM32DFU;
    } else if (vendorID == 0x1C11 && productID == 0xB007) { // Kiibohd
        deviceName = @"Kiibohd";
        deviceType = Kiibohd;
    } else if (vendorID == 0x16C0 && productID == 0x05DF) { // Objective Development BootloadHID
        deviceName = @"BootloadHID";
        deviceType = BootloadHID;
    } else if (vendorID == 0x16C0 && productID == 0x05DC) { // USBAsp and USBAspLoader
        deviceName = @"USBAsp";
        deviceType = USBAsp;
    } else if (vendorID == 0x1791 && productID == 0x0C9F) { // USB Tiny
        deviceName = @"USB Tiny";
        deviceType = USBTiny;
    } else if (vendorID == 0x1EAF && productID == 0x0003) { // STM32Duino
        deviceName = @"STM32Duino";
        deviceType = STM32Duino;
    } else {
        return;
    }

    NSString *calloutDeviceString = @"";
    NSString *mountPointString = @"";
    if (calloutDevice != nil) {
        calloutDeviceString = [NSString stringWithFormat:@" [%@]", calloutDevice];
    }
    if (mountPoint != nil) {
        mountPointString = [NSString stringWithFormat:@" [%@]", mountPoint];
    }

    [_printer print:[NSString stringWithFormat:
        @"%@ device %@: %@ %@ (%04X:%04X:%04X)%@%@",
        deviceName,
        connected ? @"connected" : @"disconnected",
        vendorString,
        productString,
        vendorID,
        productID,
        revisionBCD,
        calloutDeviceString,
        mountPointString
    ] withType:MessageType_Bootloader];

    if (connected) {
        devicesAvailable[deviceType]++;
        [delegate deviceConnected:deviceType];
    } else {
        devicesAvailable[deviceType]--;
        [delegate deviceDisconnected:deviceType];
    }
}

+ (BOOL)areDevicesAvailable {
    BOOL available = NO;
    for (int i = 0; i < NumberOfChipsets; i++) {
        available |= devicesAvailable[i];
    }
    return available;
}

+ (BOOL)canFlash:(Chipset)chipset {
    return (devicesAvailable[chipset] > 0);
}
@end
